kbase-extension/static/kbase/js/common/cellComponents/filePathWidget.js

```diff
@@ -23,21 +23,101 @@ define([
         cssBaseClass = 'kb-file-path',
         cssClassType = 'parameter';
 
+    /**
+     *
+     * @param {object} config keys:
+     *  - paramIds: Array of strings, parameter ids
+     *  - bus - the main bus that's used for communicating up to the controlling widget
+     *  - workspaceId - the id of the workspace we should use for searching for objects
+     *  - initialParams - Array of objects. Each item represents a row of parameter values.
+     *    So each item has an object with parameter id -> value
+     * @returns
+     */
     function factory(config) {
         const runtime = Runtime.make(),
+            // paramsBus is used to communicate from this parameter container to the parent that
+            // created and owns it
             paramsBus = config.bus,
             workspaceId = config.workspaceId,
             initialParams = config.initialParams,
+            paramIds = config.paramIds, // these aren't in the right order, get the right order in start()
             model = Props.make(),
+            /**
+             * Internal data model
+             * The data model / view model is a structure that maintains the overall internal
+             * state of this widget. It holds:
+             *  - all input widgets in the cell
+             *  - all values, kept in an order that's useful for the cell
+             *  - all row information, in a way that's (generally) easily accessible and easy to
+             *    update.
+             *  - each row gets its own immutible unique id, which is used for widget communication and
+             *    management
+             *  - row order, since we can delete from anywhere, can shift and wobble, so that's not helpful.
+             *    instead, we map from rowId -> row order, and update that on insertions and deletions.
+             * In general, we need to support 3 actions.
```

Comment:
off by one error?

Reply:
Not sure what you mean here - are you seeing an off by one error? Or something off in the logic here?

Reply:
"In general, we need to support 3 actions"
List of four actions 🙂

Reply:
That's, uh, zero-indexed. Yeah.

Reply:
Ok, fixed.